---
title: 'SCENIC: Introduction and setup'
subtitle: SCENIC (Single Cell rEgulatory Network Inference and Clustering)
abstract: SCENIC is a tool to infer Gene Regulatory Networks
  and their associated cell states from single-cell RNA-seq data.
  This introductory vignette provides an overview of the steps of the workflow,
  installation, and source of the dataset used as example.
output:
  BiocStyle::html_document:
    number_sections: no
    toc: yes
    css: corrected.css
  html_notebook:
    toc: yes
  pdf_document:
    toc: yes
package: '`r pkg_ver("SCENIC")`'
bibliography: references.bib
vignette: >
  %\VignetteIndexEntry{SCENIC Setup}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r libraries, echo=FALSE, message=FALSE, warning=FALSE}
# Suppress loading messages when building the HTML
suppressPackageStartupMessages({
library(GEOquery)
library(data.table)
library(Biobase)
library(SingleCellExperiment)
})

# Do not convert strings to factors
options(stringsAsFactors=FALSE)

# To build a personalized report, update this working directory:
# knitr::opts_knit$set(root.dir = 'SCENIC_MouseBrain')
```

# Introduction to SCENIC

SCENIC (Single Cell rEgulatory Network Inference and Clustering) is a tool to simultaneously reconstruct gene regulatory networks and identify stable cell states from single-cell RNA-seq data. The gene regulatory network is inferred based on co-expression and DNA motif analysis, and then the network activity is analyzed in each cell to identify the recurrent cellular states.

## More info & citation

SCENIC's approach and its application to several datasets (e.g. usage examples) was presented in the following article:

> Aibar et al. (2017) SCENIC: single-cell regulatory network inference and clustering. Nature Methods. doi: [10.1038/nmeth.4463](http://dx.doi.org/10.1038/nmeth.4463).

Please, also cite this article if you use SCENIC in your research.

# Requirements

## Species
The current version of SCENIC supports *human*, *mouse* and *fly* (Drosophila melanogaster).

To apply SCENIC to other species, it would require adjustments on the second step (e.g. new RcisTarget databases or using a diferent motif-enrichment-analysis tool).


## Input: expression matrix
The input to SCENIC is the single-cell RNA-seq **expression matrix**:

- Each column corresponds to a sample (cell) and each row corresponds to a gene.

- The gene ID should be the **gene-symbol** and stored as `rownames` (for compatibility with RcisTarget annotation databases).

- Expression **units**: The preferred expression values are gene-summarized counts (which might or might not use unique molecular identifiers, UMI).
Other measurements, such as counts or transcripts per million (TPM) and FPKM/RPKM, are also accepted as input. However, note that some authors recommend avoiding within sample normalization (i.e. TPM) for co-expression analysis (first step of SCENIC) because they may induce artificial co-variation (*@crow2016*). Although more work is still needed in order to determine the best normalization approach (*@vallejos2017*), we have applied SCENIC to datasets using raw (logged) UMI counts, normalized UMI counts, and TPM and all of them provided reliable results (see *@aibar2017*). The remaining steps of the workflow should not be affected by the expression units/normalization, as the expression is not taken into account for the motif analysis, and AUCell is within-cell ranking-based (wich works as an implicit normalization).

# Installation

SCENIC is based on three R packages:

1. `r Biocpkg("GENIE3")` to infer the co-expression network

2. `r Biocpkg("RcisTarget")` for the analysis of transcription factor binding motifs

3. `r Biocpkg("AUCell")` to identify cells with active gene sets (gene-network) in scRNA-seq data

Therefore, you will need to install these three packages, and some extra dependencies, to run SCENIC:

```{r install, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(c("GENIE3", "AUCell", "RcisTarget"), version = "3.8")
# Also required:
install.packages('zoo')

# Recommended to run AUCell:
BiocManager::install(c("mixtools", "rbokeh"))
# To visualize the binary matrices and perform t-SNEs:
BiocManager::install(c("NMF", "pheatmap", "Rtsne", "R2HTML"))
# To support paralell execution (not available in Windows):
BiocManager::install(c("doMC", "doRNG"))
# To export/visualize in http://scope.aertslab.org
if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
devtools::install_github("aertslab/SCopeLoomR", build_vignettes = TRUE)

# Other dependencies for the examples (lower priority)
BiocManager::install(c("SingleCellExperiment"))
```

AUCell and RcisTarget were updated recently (May 2018/Jan 2019). Please, make sure you have the recommended versions:
**AUCell >=1.4.1** (minimum 1.2.4), **RcisTarget>=1.2.0** (minimum 1.0.2), and **GENIE3>=1.4.0** (minimum 1.2.1).

```{r versions, eval=FALSE}
packageVersion("AUCell")
packageVersion("RcisTarget")
packageVersion("GENIE3")
```

In case you have trobule installing the packages with `biocLite` (for example, if you don't have the newest version of R), you may try installing them from Github or directly from the Bioconductor links:
```{r alternativeDownload, eval=FALSE}
# Github:
devtools::install_github("aertslab/AUCell")
devtools::install_github("aertslab/RcisTarget")
devtools::install_github("aertslab/GENIE3")
# Bioconductor
install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/AUCell_1.4.1.tar.gz", repos=NULL)
install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/RcisTarget_1.2.1.tar.gz", repos=NULL)
install.packages("https://bioconductor.org/packages/release/bioc/src/contrib/GENIE3_1.4.3.tar.gz", repos=NULL)
```

You are now ready to install SCENIC:
```{r installSCENIC, eval=FALSE}
# install.packages("devtools")
devtools::install_github("aertslab/SCENIC", ref="v1.1.0")
packageVersion("SCENIC")
```

## Species-specific databases

In addition to the R-packages, you will also need to download the species-specific databases for RcisTarget (the motif rankings). 
The links to all the available databases are available in [our website](https://resources.aertslab.org/cistarget/).
By default, SCENIC uses the databases that score the motifs in the promoter of the genes (up to 500bp upstream the TSS), and in the 20kb around the TSS (+/-10kbp).

For **human**:
```{r downloadDatabases_Human, eval=FALSE}
dbFiles <- c("https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-500bp-upstream-7species.mc9nr.feather",
"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather")
# mc9nr: Motif collection version 9: 24k motifs
```

For **mouse**:
```{r downloadDatabases_Mouse, eval=FALSE}
dbFiles <- c("https://resources.aertslab.org/cistarget/databases/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-500bp-upstream-7species.mc9nr.feather",
"https://resources.aertslab.org/cistarget/databases/mus_musculus/mm9/refseq_r45/mc9nr/gene_based/mm9-tss-centered-10kb-7species.mc9nr.feather")
# mc9nr: Motif collection version 9: 24k motifs
```

For **fly**:
```{r downloadDatabases_Fly, eval=FALSE}
dbFiles <- c("https://resources.aertslab.org/cistarget/databases/drosophila_melanogaster/dm6/flybase_r6.02/mc8nr/gene_based/dm6-5kb-upstream-full-tx-11species.mc8nr.feather")
# mc8nr: Motif collection version 8: 20k motifs
```

Download the *.feather* files (and *.descr*, if available) for the relevant organism. [Aprox file size: 1GB]

*To avoid problems or incomplete downloads (especially if you have a slow connection), we recommend to use [zsync_curl](https://resources.aertslab.org/cistarget/help.html). If you prefer to download directly from R, you can try the following code:*
```{r downloadDatabases, eval=FALSE}
# dir.create("cisTarget_databases"); setwd("cisTarget_databases") # if needed
for(featherURL in dbFiles)
{
  download.file(featherURL, destfile=basename(featherURL)) # saved in current dir
  descrURL <- gsub(".feather$", ".descr", featherURL)
  if(file.exists(descrURL)) download.file(descrURL, destfile=basename(descrURL))
}
```

To confirm that the databases were downloaded correctly, we recommend to confirm its `sha256sum`: https://resources.aertslab.org/cistarget/databases/sha256sum.txt 

After these setup steps, SCENIC is ready to run! To start, see the vignette "SCENIC_Running".

# Some tips...

## Template for your own analysis
You can use the [R notebooks](http://rmarkdown.rstudio.com/r_notebooks.html) of this workflow as template for your own data (i.e. copy the .Rmd file, and edit it in RStudio).
```{r template, eval=FALSE}
vignetteFile <- file.path(system.file('doc', package='SCENIC'), "SCENIC_Running.Rmd")
file.copy(vignetteFile, "SCENIC_myRun.Rmd")
# or: 
vignetteFile <- "https://raw.githubusercontent.com/aertslab/SCENIC/master/vignettes/SCENIC_Running.Rmd"
download.file(vignetteFile, "SCENIC_myRun.Rmd")
```

Note that some steps of this workflow take considerable time. To avoid re-running these steps when *knitting* the vignette (i.e. create the HTML report), we have added `eval=FALSE` to some code chunks and `load()` its output in the next. Feel free to adapt these to your needs.

## Help
At any time, you an access the help for any function used in this workflow (i.e. for details on their arguments), and the vignettes of the other steps of the workflow with the following commands:

```{r help, eval=FALSE}
## Get help for a function:
?runSCENIC_3_scoreCells
help(runSCENIC_3_scoreCells) # equivalent

## See the available tutorials:
vignette(package="SCENIC") # list
vignette("SCENIC_Running") # open
```

# Sample dataset: Download & format

This tutorial is run on a real dataset to illustrate the results obtained at each step. However, keep in mind that some steps of the workflow (i.e. GENIE3, motif enrichment) can take hours to run. Feel free to use your own dataset instead.

The output files from this example are available at http://scenic.aertslab.org/examples/ 

As example, we will use a dataset containing 3005 cells from the adult mouse brain, including neurons (e.g. pyramidal neurons and interneurons) and glia (oligodendrocytes, astrocytes/ependymal, endothelial/mural and microglia).

> Zeisel, A., et al. (2015). Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq. Science 347, 1138â€“1142. doi: [10.1126/science.aaa1934](http://dx.doi.org/10.1126/science.aaa1934)

The expression matrix can be downloaded from GEO accession number GSE60361 or the author's website: http://linnarssonlab.org/cortex/.
The expression is provided as Unique Molecular Identifier counts:

```{r GEOdownload, results='hide', message=FALSE, eval=FALSE}
# dir.create("SCENIC_MouseBrain"); setwd("SCENIC_MouseBrain") # if needed

# (This may take a few minutes)
library(GEOquery)
geoFile <- getGEOSuppFiles("GSE60361", makeDirectory=FALSE)
gzFile <- grep("Expression", basename(rownames(geoFile)), value=TRUE)
txtFile <- gsub(".gz", "", gzFile)
gunzip(gzFile, destname=txtFile, remove=TRUE)

library(data.table)
geoData <- fread(txtFile, sep="\t")
geneNames <- unname(unlist(geoData[,1, with=FALSE]))
exprMatrix <- as.matrix(geoData[,-1, with=FALSE])
rm(geoData)
dim(exprMatrix)
rownames(exprMatrix) <- geneNames
exprMatrix[1:5,1:4]

# Remove file downloaded:
file.remove(txtFile)
```

The cell types are available at the author's website. For simplicity, here we will load them from AUCell package:
```{r cellLabels}
cellLabels <- paste(file.path(system.file('examples', package='AUCell')), "mouseBrain_cellLabels.tsv", sep="/")
cellLabels <- read.table(cellLabels, row.names=1, header=TRUE, sep="\t")
cellLabels <- as.data.frame(cellLabels)
colnames(cellLabels) <- "CellType"
```

Finally, this information can be merged into a single Bioconductor object.
Here we use a `SingleCellExperiment`, but a `SummarizedExperiment`, `ExpressionSet` or similar classes would also be valid. You might just need to adjust the accessor functions `colData()`, and `counts()` in the next steps.
```{r createEset, warning=FALSE, eval=FALSE}
# TODO: Change to loom file
exprMatrix <- exprMatrix[unique(rownames(exprMatrix)),] # Remove duplicated rows
dim(exprMatrix)

SingleCellExperiment
library(SingleCellExperiment)
sceMouseBrain <- SingleCellExperiment(assays = list(counts = exprMatrix),
colData=data.frame(cellLabels[colnames(exprMatrix),, drop=FALSE]))

# setwd("SCENIC_MouseBrain")
dir.create("data")
saveRDS(sceMouseBrain, file="data/sceMouseBrain.Rds")
```

# References
